### PR TITLE
Simplify boot UART and reset interaction

### PR DESCRIPTION
**Pull request by r1mikey**
The boot UART is presently a hardcoded, platform specific chunk of code that is needed very early in boot. This all prevents us from introducing platmods for platform-specifics.

Refactor bootsvcs UART and reset interactions into a separate file and drive the selection of the UART implementation from xboot_info. This means that inetboot (already deeply platform specific) is now responsible for setting up the correct parameters for the UART and for performing any initialisation required, as the bsvc code in the kernel only knows how to send and receive characters.

### CODE REVIEW

**Comment by r1mikey** [comment]
Putting this up now to get early feedback.

Tested on qemu virt/FDT: https://gist.github.com/r1mikey/a729746a480fdb225efcf7619ff02f8c

Still need to test on Raspberry Pi.

@richlowe  Do you think it's time to delete Meson GXBB? I don't think anybody is running it.

**Comment by richlowe** [comment]
I think deleting GXBB is fine.  I keep going to do it and then not in my current far-too-large wad of changes toward PCIe.

I see the logic of the UART changes.  Is this what other systems do to get an early uart? (my dream still being to be able to use a lightly modified u-boot one day, with no inetboot)

**Comment by r1mikey** [comment]
> I think deleting GXBB is fine. I keep going to do it and then not in my current far-too-large wad of changes toward PCIe.

I'll get this done.

> I see the logic of the UART changes. Is this what other systems do to get an early uart?

FreeBSD (I think - it might be NetBSD)) has a compile-time option to hardcode the base address of the early UART (this is not great).
Linux passes earlycon, which specifies the type, base and speed. On aarch64 you can just pass earlycon (as a boolean) for ACPI systems and Linux will choose the SPCR or DBG2 UARTs.
Windows uses DBG2 and SPCR as well.

> (my dream still being to be able to use a lightly modified u-boot one day, with no inetboot)

This is exactly what I'd like to see. In the FDT case we should be able to have one inetboot, using compatible on the root of the FDT to select board-specific code where needed. This would set up "firmware agnostic" xboot_info, then hand off to unix, which would do similar things to dispatch minimal board-specific code (perhaps via platmod).

In the ACPI/UEFI case we'd do exactly the same thing, preparing firmware agnostic xboot_info and handing off to unix. 

The FDT and ACPI/UEFI paths could, in theory, even be the same inetboot, examining how they were booted to select the firmware interfaces to use for setup. If this doesn't get too large we'd end up with one inetboot, which is much easier to reason about.

With all of that done, moving a minimised inetboot into a dboot-like self-relocating blob becomes feasible. The only thing that I'm stuck on here is finding the real unix entry point in the FDT case. The dboot case is fairly simple, as loader64.efi passes the unix ELF symbol table through to dboot as a "module". I'll need to figure out if uboot could do something similar (perhaps via the FDT).

**Comment by r1mikey** [comment]
Tested on Raspberry Pi: https://gist.github.com/r1mikey/ee8e82d5844ff8a5dff5c24313fda75c

The following lines are printed via `_kobj_printf`, which uses bsvc character printing:
```
module /platform/RaspberryPi,4/kernel/aarch64/unix: text at [0xfffffffffe000000, 0xfffffffffe115057] data at 0xfffffffffe116000
module /kernel/aarch64/genunix: text at [0xfffffffffe168a20, 0xfffffffffe4327ef] data at 0xfffffffffe432800
```

**Comment by r1mikey** [comment]
I have the meson-gxbb removal stacked on top of this change - will rebase and put up a PR once this is in.

File: usr/src/uts/aarch64/sys/bootinfo.h [hunk 1]
```diff
@@ -38,6 +38,12 @@ extern "C" {
 
 #define	MAX_BOOT_MODULES	99
 
+#define	XBI_BSVC_UART_PL011	0x0003
```
Comment (richlowe) [inline usr/src/uts/aarch64/sys/bootinfo.h]: enum? maybe?

**Review by richlowe — APPROVED** [review]
i only have the one, incredibly tiny and optional comment.

You said you were putting this up early.  Does that mean you don't want me to merge it yet?

**Comment by r1mikey** [comment]
> You said you were putting this up early. Does that mean you don't want me to merge it yet?

I'll do the enum change, then it should be good to go.

**Comment by r1mikey** [comment]
Should be good merge now.

**Comment by richlowe** [comment]
I pushed this, with some cstyle fixes because of https://www.illumos.org/issues/16888. 
Thanks

**Pull request closed without merging**